### PR TITLE
compiler: adds support for FunctionListeners

### PR DESCRIPTION
**Pull request by mathetake**
Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

File: internal/engine/compiler/engine.go [hunk 1]
```diff
@@ -455,15 +464,21 @@ func (e *engine) CompileModule(ctx context.Context, module *wasm.Module) error {
 
 	importedFuncs := module.ImportFuncCount()
 	funcs := make([]*code, len(module.FunctionSection))
+	ln := len(listeners)
 	for i, ir := range irs {
+		var withListener bool
+		if i < ln {
+			withListener = listeners[i] != nil
+		}
+
```
Comment (mathetake) [inline internal/engine/compiler/engine.go:473]: this makes it possible for us to assemble machine code on the function listeners calling only when a function has non-nil listeners, and in the machine code generation, we call the listener on preamble and function return as opposed to the case of interpreter where it's calling it on the caller side.

**Comment by mathetake** [comment]
ok switched all tests using listener to compilers, and passed all of them on arm64!

**Comment by mathetake** [comment]
passed all tests both on amd64 and arm64!

File: experimental/listener_example_test.go [hunk 2]
```diff
@@ -58,7 +58,7 @@ func Example_customListenerFactory() {
 	// Set context to one that has an experimental listener
 	ctx := context.WithValue(context.Background(), FunctionListenerFactoryKey{}, u)
 
-	r := wazero.NewRuntimeWithConfig(ctx, wazero.NewRuntimeConfigInterpreter())
+	r := wazero.NewRuntime(ctx)
```
Comment (codefromthecrypt) [inline experimental/listener_example_test.go:61]: yay

File: internal/engine/compiler/engine.go [hunk 3]
```diff
@@ -118,6 +119,20 @@ type (
 
 		// initialFn is the initial function for this call engine.
 		initialFn *function
+
+		// ctx is the context.Context passed to all the host function calls.
+		// This is modified when there's a function listener call, otherwise it's always the context.Context
+		// passed to the Call API.
+		ctx context.Context
+		// contextStack is a stack of contexts which is pushed and popped by function listeners.
+		// This is used and modified when there's function listeners.
```
Comment (codefromthecrypt) [inline internal/engine/compiler/engine.go:128]:
```suggestion
		// This is used and modified when there are function listeners.
```

**Review by codefromthecrypt — APPROVED** [review]
looking good. we might need some doc clarification.

So, I think listeners need to be in context for compile. In the compile engine, must the count match at runtime, or is the listener value only read at compile time and re-used thereafter?

Can we can tighten it up to only reading the listeners during compile and that's it, ex proved by passing context.Background to instantiate and also compile? If not should we or how much effort would take to do that?

**Comment by codefromthecrypt** [comment]
clarification: looks like this is compile config, so we can add a test to prove it and clarify on the context key that it applies to CompileModule or things that imply that. ignore my other comment otherwise!

**Review by codefromthecrypt — APPROVED** [review]
coolio

**Pull request merged**